File: src/ads/structures.py
"""Regular Python data-structure tier: simple, readable implementations.

Mirror the Java / C++ / Rust ports. Each structure exposes the operation
contracts listed in ``catalog/problems.json``.
"""

from __future__ import annotations

from collections import deque as _deque

_SENTINEL = -1


class Stack:
    """LIFO stack backed by a list."""

    __slots__ = ("_items",)

    def __init__(self) -> None:
        self._items: list[int] = []

    def push(self, value: int) -> None:
        self._items.append(value)

    def pop(self) -> int:
        return self._items.pop() if self._items else _SENTINEL

    def peek(self) -> int:
        return self._items[-1] if self._items else _SENTINEL

    def is_empty(self) -> bool:
        return not self._items

    def size(self) -> int:
        return len(self._items)


class Queue:
    """FIFO queue backed by a deque."""

    __slots__ = ("_items",)

    def __init__(self) -> None:
        self._items: _deque[int] = _deque()

    def enqueue(self, value: int) -> None:
        self._items.append(value)

    def dequeue(self) -> int:
        return self._items.popleft() if self._items else _SENTINEL

    def peek(self) -> int:
        return self._items[0] if self._items else _SENTINEL

    def is_empty(self) -> bool:
        return not self._items

    def size(self) -> int:
        return len(self._items)


class _Node:
    __slots__ = ("next", "value")

    def __init__(self, value: int) -> None:
        self.value = value
        self.next: _Node | None = None


class LinkedList:
    """Singly linked list."""

    __slots__ = ("_head", "_size", "_tail")

    def __init__(self) -> None:
        self._head: _Node | None = None
        self._tail: _Node | None = None
        self._size = 0

    def append(self, value: int) -> None:
        node = _Node(value)
        if self._tail is None:
            self._head = self._tail = node
        else:
            self._tail.next = node
            self._tail = node
        self._size += 1

    def prepend(self, value: int) -> None:
        node = _Node(value)
        node.next = self._head
        self._head = node
        if self._tail is None:
            self._tail = node
        self._size += 1

    def get(self, index: int) -> int:
        if index < 0 or index >= self._size:
            return _SENTINEL
        node = self._head
        for _ in range(index):
            node = node.next  # type: ignore[union-attr]
        return node.value  # type: ignore[union-attr]

    def contains(self, value: int) -> bool:
        node = self._head
        while node is not None:
            if node.value == value:
                return True
            node = node.next
        return False

    def remove(self, value: int) -> bool:
        node, previous = self._head, None
        while node is not None:
            if node.value == value:
                if previous is None:
                    self._head = node.next
                else:
                    previous.next = node.next
                if node is self._tail:
                    self._tail = previous
                self._size -= 1
                return True
            previous, node = node, node.next
        return False

    def size(self) -> int:
        return self._size

    def to_list(self) -> list[int]:
        out: list[int] = []
        node = self._head
        while node is not None:
            out.append(node.value)
            node = node.next
        return out


class BST:
    """Binary search tree."""

    __slots__ = ("_root", "_size")

    def __init__(self) -> None:
        self._root: _BSTNode | None = None
        self._size = 0

    def insert(self, value: int) -> None:
        if self._root is None:
            self._root = _BSTNode(value)
            self._size = 1
            return
        node = self._root
        while True:
            if value < node.value:
                if node.left is None:
                    node.left = _BSTNode(value)
                    break
                node = node.left
            elif value > node.value:
                if node.right is None:
                    node.right = _BSTNode(value)
                    break
                node = node.right
            else:
                return
        self._size += 1

    def contains(self, value: int) -> bool:
        node = self._root
        while node is not None:
            if value == node.value:
                return True
            node = node.left if value < node.value else node.right
        return False

    def min(self) -> int:
        node = self._root
        if node is None:
            return _SENTINEL
        while node.left is not None:
            node = node.left
        return node.value

    def max(self) -> int:
        node = self._root
        if node is None:
            return _SENTINEL
        while node.right is not None:
            node = node.right
        return node.value

    def height(self) -> int:
        return _height(self._root)

    def size(self) -> int:
        return self._size

    def in_order(self) -> list[int]:
        out: list[int] = []
        _in_order(self._root, out)
        return out

    def remove(self, value: int) -> bool:
        found = self._contains(self._root, value)
        if not found:
            return False
        self._root = _remove(self._root, value)
        self._size -= 1
        return True

    @staticmethod
    def _contains(node: _BSTNode | None, value: int) -> bool:
        if node is None:
            return False
        if value == node.value:
            return True
        return BST._contains(node.left if value < node.value else node.right, value)


class _BSTNode:
    __slots__ = ("left", "right", "value")

    def __init__(self, value: int) -> None:
        self.value = value
        self.left: _BSTNode | None = None
        self.right: _BSTNode | None = None


def _height(node: _BSTNode | None) -> int:
    if node is None:
        return -1
    return 1 + max(_height(node.left), _height(node.right))


def _in_order(node: _BSTNode | None, out: list[int]) -> None:
    if node is None:
        return
    _in_order(node.left, out)
    out.append(node.value)
    _in_order(node.right, out)


def _remove(node: _BSTNode | None, value: int) -> _BSTNode | None:
    if node is None:
        return None
    if value < node.value:
        node.left = _remove(node.left, value)
    elif value > node.value:
        node.right = _remove(node.right, value)
    else:
        if node.left is None:
            return node.right
        if node.right is None:
            return node.left
        successor = node.right
        while successor.left is not None:
            successor = successor.left
        node.value = successor.value
        node.right = _remove(node.right, successor.value)
    return node


class Trie:
    """Prefix trie storing unique words."""

    __slots__ = ("_root", "_size")

    def __init__(self) -> None:
        self._root: _TrieNode = _TrieNode()
        self._size = 0

    def insert(self, word: str) -> None:
        node = self._root
        for ch in word:
            if ch not in node.children:
                node.children[ch] = _TrieNode()
            node = node.children[ch]
        if not node.terminal:
            node.terminal = True
            self._size += 1

    def search(self, word: str) -> bool:
        node = self._root
        for ch in word:
            if ch not in node.children:
                return False
            node = node.children[ch]
        return node.terminal

    def starts_with(self, prefix: str) -> bool:
        node = self._root
        for ch in prefix:
            if ch not in node.children:
                return False
            node = node.children[ch]
        return True

    def size(self) -> int:
        return self._size


class _TrieNode:
    __slots__ = ("children", "terminal")

    def __init__(self) -> None:
        self.children: dict[str, _TrieNode] = {}
        self.terminal = False


class MinHeap:
    """Binary min-heap over a list."""

    __slots__ = ("_items",)

    def __init__(self) -> None:
        self._items: list[int] = []

    def push(self, value: int) -> None:
        self._items.append(value)
        self._sift_up(len(self._items) - 1)

    def pop(self) -> int:
        if not self._items:
            return _SENTINEL
        top = self._items[0]
        last = self._items.pop()
        if self._items:
            self._items[0] = last
            self._sift_down(0)
        return top

    def peek(self) -> int:
        return self._items[0] if self._items else _SENTINEL

    def is_empty(self) -> bool:
        return not self._items

    def size(self) -> int:
        return len(self._items)

    def _sift_up(self, index: int) -> None:
        while index > 0:
            parent = (index - 1) // 2
            if self._items[parent] <= self._items[index]:
                return
            self._items[parent], self._items[index] = (
                self._items[index],
                self._items[parent],
            )
            index = parent

    def _sift_down(self, index: int) -> None:
        size = len(self._items)
        while True:
            left = 2 * index + 1
            right = left + 1
            smallest = index
            if left < size and self._items[left] < self._items[smallest]:
                smallest = left
            if right < size and self._items[right] < self._items[smallest]:
                smallest = right
            if smallest == index:
                return
            self._items[index], self._items[smallest] = (
                self._items[smallest],
                self._items[index],
            )
            index = smallest

File: languages/java/src/ads/Structures.java
package ads;

import java.util.ArrayList;
import java.util.List;

/**
 * Data-structure implementations (Java).
 *
 * Operation contracts match catalog/problems.json. Sentinel value for
 * missing/empty reads is -1 and is used consistently across all languages.
 */
public final class Structures {

    private Structures() {
    }

    /** LIFO stack backed by an array list. */
    public static final class Stack {
        private final List<Integer> items = new ArrayList<>();

        public void push(int value) {
            items.add(value);
        }

        public int pop() {
            return items.isEmpty() ? -1 : items.remove(items.size() - 1);
        }

        public int peek() {
            return items.isEmpty() ? -1 : items.get(items.size() - 1);
        }

        public boolean isEmpty() {
            return items.isEmpty();
        }

        public int size() {
            return items.size();
        }
    }

    /** FIFO queue backed by an array list ring offset. */
    public static final class Queue {
        private final List<Integer> items = new ArrayList<>();
        private int head;

        public void enqueue(int value) {
            items.add(value);
        }

        public int dequeue() {
            if (size() == 0) {
                return -1;
            }
            return items.get(head++);
        }

        public int peek() {
            return size() == 0 ? -1 : items.get(head);
        }

        public boolean isEmpty() {
            return size() == 0;
        }

        public int size() {
            return items.size() - head;
        }
    }

    /** Singly linked list. */
    public static final class LinkedList {
        private static final class Node {
            int value;
            Node next;

            Node(int value) {
                this.value = value;
            }
        }

        private Node head;
        private Node tail;
        private int count;

        public void append(int value) {
            Node node = new Node(value);
            if (tail == null) {
                head = tail = node;
            } else {
                tail.next = node;
                tail = node;
            }
            count++;
        }

        public void prepend(int value) {
            Node node = new Node(value);
            node.next = head;
            head = node;
            if (tail == null) {
                tail = node;
            }
            count++;
        }

        public int get(int index) {
            if (index < 0 || index >= count) {
                return -1;
            }
            Node node = head;
            for (int i = 0; i < index; i++) {
                node = node.next;
            }
            return node == null ? -1 : node.value;
        }

        public boolean contains(int value) {
            Node node = head;
            while (node != null) {
                if (node.value == value) {
                    return true;
                }
                node = node.next;
            }
            return false;
        }

        public boolean remove(int value) {
            Node previous = null;
            Node node = head;
            while (node != null) {
                if (node.value == value) {
                    if (previous == null) {
                        head = node.next;
                    } else {
                        previous.next = node.next;
                    }
                    if (node == tail) {
                        tail = previous;
                    }
                    count--;
                    return true;
                }
                previous = node;
                node = node.next;
            }
            return false;
        }

        public int size() {
            return count;
        }

        public List<Integer> toList() {
            List<Integer> out = new ArrayList<>();
            Node node = head;
            while (node != null) {
                out.add(node.value);
                node = node.next;
            }
            return out;
        }
    }

    /** Binary search tree. */
    public static final class BST {
        private static final class Node {
            int value;
            Node left;
            Node right;

            Node(int value) {
                this.value = value;
            }
        }

        private Node root;
        private int count;

        public void insert(int value) {
            if (root == null) {
                root = new Node(value);
                count = 1;
                return;
            }
            Node node = root;
            while (true) {
                if (value < node.value) {
                    if (node.left == null) {
                        node.left = new Node(value);
                        break;
                    }
                    node = node.left;
                } else if (value > node.value) {
                    if (node.right == null) {
                        node.right = new Node(value);
                        break;
                    }
                    node = node.right;
                } else {
                    return;
                }
            }
            count++;
        }

        public boolean contains(int value) {
            Node node = root;
            while (node != null) {
                if (value == node.value) {
                    return true;
                }
                node = value < node.value ? node.left : node.right;
            }
            return false;
        }

        public int min() {
            Node node = root;
            if (node == null) {
                return -1;
            }
            while (node.left != null) {
                node = node.left;
            }
            return node.value;
        }

        public int max() {
            Node node = root;
            if (node == null) {
                return -1;
            }
            while (node.right != null) {
                node = node.right;
            }
            return node.value;
        }

        public int height() {
            return heightOf(root);
        }

        private static int heightOf(Node node) {
            if (node == null) {
                return -1;
            }
            return 1 + Math.max(heightOf(node.left), heightOf(node.right));
        }

        public int size() {
            return count;
        }

        public List<Integer> inOrder() {
            List<Integer> out = new ArrayList<>();
            inOrderOf(root, out);
            return out;
        }

        private static void inOrderOf(Node node, List<Integer> out) {
            if (node == null) {
                return;
            }
            inOrderOf(node.left, out);
            out.add(node.value);
            inOrderOf(node.right, out);
        }

        public boolean remove(int value) {
            if (!contains(value)) {
                return false;
            }
            root = removeNode(root, value);
            count--;
            return true;
        }

        private static Node removeNode(Node node, int value) {
            if (node == null) {
                return null;
            }
            if (value < node.value) {
                node.left = removeNode(node.left, value);
            } else if (value > node.value) {
                node.right = removeNode(node.right, value);
            } else {
                if (node.left == null) {
                    return node.right;
                }
                if (node.right == null) {
                    return node.left;
                }
                Node successor = node.right;
                while (successor.left != null) {
                    successor = successor.left;
                }
                node.value = successor.value;
                node.right = removeNode(node.right, successor.value);
            }
            return node;
        }
    }

    /** Prefix trie. */
    public static final class Trie {
        private static final class Node {
            final java.util.HashMap<Character, Node> children = new java.util.HashMap<>();
            boolean terminal;
        }

        private final Node root = new Node();
        private int count;

        public void insert(String word) {
            Node node = root;
            for (int i = 0; i < word.length(); i++) {
                char ch = word.charAt(i);
                Node child = node.children.get(ch);
                if (child == null) {
                    child = new Node();
                    node.children.put(ch, child);
                }
                node = child;
            }
            if (!node.terminal) {
                node.terminal = true;
                count++;
            }
        }

        public boolean search(String word) {
            Node node = root;
            for (int i = 0; i < word.length(); i++) {
                Node child = node.children.get(word.charAt(i));
                if (child == null) {
                    return false;
                }
                node = child;
            }
            return node.terminal;
        }

        public boolean startsWith(String prefix) {
            Node node = root;
            for (int i = 0; i < prefix.length(); i++) {
                Node child = node.children.get(prefix.charAt(i));
                if (child == null) {
                    return false;
                }
                node = child;
            }
            return true;
        }

        public int size() {
            return count;
        }
    }

    /** Binary min-heap. */
    public static final class MinHeap {
        private final List<Integer> items = new ArrayList<>();

        public void push(int value) {
            items.add(value);
            siftUp(items.size() - 1);
        }

        public int pop() {
            if (items.isEmpty()) {
                return -1;
            }
            int top = items.get(0);
            int last = items.remove(items.size() - 1);
            if (!items.isEmpty()) {
                items.set(0, last);
                siftDown(0);
            }
            return top;
        }

        public int peek() {
            return items.isEmpty() ? -1 : items.get(0);
        }

        public boolean isEmpty() {
            return items.isEmpty();
        }

        public int size() {
            return items.size();
        }

        private void siftUp(int index) {
            while (index > 0) {
                int parent = (index - 1) / 2;
                if (items.get(parent) <= items.get(index)) {
                    return;
                }
                swap(parent, index);
                index = parent;
            }
        }

        private void siftDown(int index) {
            int size = items.size();
            while (true) {
                int left = 2 * index + 1;
                int right = left + 1;
                int smallest = index;
                if (left < size && items.get(left) < items.get(smallest)) {
                    smallest = left;
                }
                if (right < size && items.get(right) < items.get(smallest)) {
                    smallest = right;
                }
                if (smallest == index) {
                    return;
                }
                swap(index, smallest);
                index = smallest;
            }
        }

        private void swap(int i, int j) {
            int tmp = items.get(i);
            items.set(i, items.get(j));
            items.set(j, tmp);
        }
    }
}